File: code/packages/typescript/aes-modes/tests/aes-modes.test.ts
/**
 * Tests for @coding-adventures/aes-modes
 *
 * Uses NIST test vectors from:
 *   - SP 800-38A (ECB, CBC, CTR)
 *   - GCM specification (GCM)
 *
 * NIST test vectors are the gold standard for verifying AES implementations.
 * Each vector provides known input/output pairs that any correct implementation
 * must reproduce exactly.
 */

import { describe, it, expect } from "vitest";
import {
  ecbEncrypt,
  ecbDecrypt,
  cbcEncrypt,
  cbcDecrypt,
  ctrEncrypt,
  ctrDecrypt,
  gcmEncrypt,
  gcmDecrypt,
  pkcs7Pad,
  pkcs7Unpad,
  fromHex,
  toHex,
} from "../src/index";

// ─────────────────────────────────────────────────────────────────────────────
// Shared test data — NIST SP 800-38A, Appendix F
// ─────────────────────────────────────────────────────────────────────────────

/** AES-128 key used across most NIST SP 800-38A test vectors. */
const NIST_KEY = fromHex("2b7e151628aed2a6abf7158809cf4f3c");

/** The first plaintext block from NIST SP 800-38A (used in ECB, CBC, CTR). */
const NIST_PLAINTEXT_BLOCK1 = fromHex("6bc1bee22e409f96e93d7e117393172a");

/** All four plaintext blocks from NIST SP 800-38A. */
const NIST_PLAINTEXT_ALL = fromHex(
  "6bc1bee22e409f96e93d7e117393172a" +
  "ae2d8a571e03ac9c9eb76fac45af8e51" +
  "30c81c46a35ce411e5fbc1191a0a52ef" +
  "f69f2445df4f9b17ad2b417be66c3710"
);

// ─────────────────────────────────────────────────────────────────────────────
// PKCS#7 Padding Tests
// ─────────────────────────────────────────────────────────────────────────────

describe("PKCS#7 Padding", () => {
  it("pads a block-aligned input with a full block of 0x10", () => {
    // When the input is already 16 bytes, we must add a full padding block.
    // This ensures unpadding always works — without this rule, there would
    // be ambiguity when the last byte of plaintext happens to be 0x01-0x10.
    const input = new Uint8Array(16).fill(0xaa);
    const padded = pkcs7Pad(input);
    expect(padded.length).toBe(32);
    // Last 16 bytes should all be 0x10 (= 16)
    for (let i = 16; i < 32; i++) {
      expect(padded[i]).toBe(16);
    }
  });

  it("pads a 13-byte input with 3 bytes of 0x03", () => {
    const input = new Uint8Array(13).fill(0xbb);
    const padded = pkcs7Pad(input);
    expect(padded.length).toBe(16);
    expect(padded[13]).toBe(3);
    expect(padded[14]).toBe(3);
    expect(padded[15]).toBe(3);
  });

  it("pads an empty input with a full block of 0x10", () => {
    const padded = pkcs7Pad(new Uint8Array(0));
    expect(padded.length).toBe(16);
    for (let i = 0; i < 16; i++) {
      expect(padded[i]).toBe(16);
    }
  });

  it("round-trips: unpad(pad(x)) === x", () => {
    const input = new Uint8Array([1, 2, 3, 4, 5]);
    const result = pkcs7Unpad(pkcs7Pad(input));
    expect(toHex(result)).toBe(toHex(input));
  });

  it("rejects invalid padding value 0", () => {
    const bad = new Uint8Array(16);
    bad[15] = 0;
    expect(() => pkcs7Unpad(bad)).toThrow("Invalid PKCS#7 padding");
  });

  it("rejects inconsistent padding bytes", () => {
    const bad = new Uint8Array(16);
    bad[15] = 2;
    bad[14] = 3; // Should be 2
    expect(() => pkcs7Unpad(bad)).toThrow("Invalid PKCS#7 padding");
  });

  it("rejects non-multiple-of-16 input", () => {
    expect(() => pkcs7Unpad(new Uint8Array(15))).toThrow("positive multiple of 16");
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// ECB Mode Tests
// ─────────────────────────────────────────────────────────────────────────────

describe("ECB Mode", () => {
  it("encrypts NIST SP 800-38A block 1 correctly", () => {
    // NIST expected: 3ad77bb40d7a3660a89ecaf32466ef97
    // Note: ecbEncrypt applies PKCS#7 padding, so the ciphertext will be
    // 32 bytes (the original 16-byte block + 16 bytes of padding encrypted).
    // We check the first block only.
    const ct = ecbEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY);
    expect(toHex(ct.slice(0, 16))).toBe("3ad77bb40d7a3660a89ecaf32466ef97");
  });

  it("encrypts all 4 NIST blocks correctly", () => {
    const ct = ecbEncrypt(NIST_PLAINTEXT_ALL, NIST_KEY);
    // 4 blocks of plaintext + 1 block of padding = 5 blocks = 80 bytes
    expect(ct.length).toBe(80);
    expect(toHex(ct.slice(0, 16))).toBe("3ad77bb40d7a3660a89ecaf32466ef97");
    expect(toHex(ct.slice(16, 32))).toBe("f5d3d58503b9699de785895a96fdbaaf");
    expect(toHex(ct.slice(32, 48))).toBe("43b1cd7f598ece23881b00e3ed030688");
    expect(toHex(ct.slice(48, 64))).toBe("7b0c785e27e8ad3f8223207104725dd4");
  });

  it("round-trips: decrypt(encrypt(x)) === x", () => {
    const pt = ecbDecrypt(ecbEncrypt(NIST_PLAINTEXT_ALL, NIST_KEY), NIST_KEY);
    expect(toHex(pt)).toBe(toHex(NIST_PLAINTEXT_ALL));
  });

  it("round-trips non-aligned plaintext", () => {
    const pt = new Uint8Array([0xde, 0xad, 0xbe, 0xef]);
    const result = ecbDecrypt(ecbEncrypt(pt, NIST_KEY), NIST_KEY);
    expect(toHex(result)).toBe("deadbeef");
  });

  it("rejects invalid ciphertext length", () => {
    expect(() => ecbDecrypt(new Uint8Array(15), NIST_KEY)).toThrow();
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// CBC Mode Tests
// ─────────────────────────────────────────────────────────────────────────────

describe("CBC Mode", () => {
  const CBC_IV = fromHex("000102030405060708090a0b0c0d0e0f");

  it("encrypts NIST SP 800-38A block 1 correctly", () => {
    // NIST expected: 7649abac8119b246cee98e9b12e9197d
    const ct = cbcEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, CBC_IV);
    expect(toHex(ct.slice(0, 16))).toBe("7649abac8119b246cee98e9b12e9197d");
  });

  it("encrypts all 4 NIST blocks correctly", () => {
    const ct = cbcEncrypt(NIST_PLAINTEXT_ALL, NIST_KEY, CBC_IV);
    expect(ct.length).toBe(80); // 64 + 16 padding
    expect(toHex(ct.slice(0, 16))).toBe("7649abac8119b246cee98e9b12e9197d");
    expect(toHex(ct.slice(16, 32))).toBe("5086cb9b507219ee95db113a917678b2");
    expect(toHex(ct.slice(32, 48))).toBe("73bed6b8e3c1743b7116e69e22229516");
    expect(toHex(ct.slice(48, 64))).toBe("3ff1caa1681fac09120eca307586e1a7");
  });

  it("round-trips: decrypt(encrypt(x)) === x", () => {
    const pt = cbcDecrypt(cbcEncrypt(NIST_PLAINTEXT_ALL, NIST_KEY, CBC_IV), NIST_KEY, CBC_IV);
    expect(toHex(pt)).toBe(toHex(NIST_PLAINTEXT_ALL));
  });

  it("round-trips non-aligned plaintext", () => {
    const pt = new Uint8Array([0xca, 0xfe, 0xba, 0xbe]);
    const result = cbcDecrypt(cbcEncrypt(pt, NIST_KEY, CBC_IV), NIST_KEY, CBC_IV);
    expect(toHex(result)).toBe("cafebabe");
  });

  it("rejects wrong IV length", () => {
    expect(() => cbcEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, new Uint8Array(8))).toThrow("16 bytes");
  });

  it("different IVs produce different ciphertext", () => {
    const iv1 = fromHex("00000000000000000000000000000000");
    const iv2 = fromHex("ffffffffffffffffffffffffffffffff");
    const ct1 = cbcEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, iv1);
    const ct2 = cbcEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, iv2);
    expect(toHex(ct1)).not.toBe(toHex(ct2));
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// CTR Mode Tests
// ─────────────────────────────────────────────────────────────────────────────

describe("CTR Mode", () => {
  // NIST SP 800-38A CTR test uses a full 16-byte IV, but our implementation
  // uses 12-byte nonce + 4-byte counter. The NIST vector:
  //   IV = f0f1f2f3 f4f5f6f7 f8f9fafb fcfdfeff
  // maps to nonce = f0f1f2f3f4f5f6f7f8f9fafb, counter starting at fcfdfeff.
  // However, our counter starts at 1, so we need to use a nonce that produces
  // the correct first counter block.
  //
  // For the NIST vector, the first counter block is:
  //   f0f1f2f3 f4f5f6f7 f8f9fafb fcfdfeff
  // Our format: nonce(12) || counter(4), with counter = 1:
  //   nonce || 00000001
  // To match NIST, we'd need nonce = f0f1f2f3f4f5f6f7f8f9fafb and counter = fcfdfeff.
  // Since our counter starts at 1 (not fcfdfeff), we'll verify the round-trip
  // and use a separate test for the NIST vector with a custom approach.

  it("round-trips: decrypt(encrypt(x)) === x", () => {
    const nonce = fromHex("f0f1f2f3f4f5f6f7f8f9fafb");
    const pt = ctrDecrypt(ctrEncrypt(NIST_PLAINTEXT_ALL, NIST_KEY, nonce), NIST_KEY, nonce);
    expect(toHex(pt)).toBe(toHex(NIST_PLAINTEXT_ALL));
  });

  it("handles non-aligned plaintext (no padding needed)", () => {
    const nonce = fromHex("000000000000000000000000");
    const pt = new Uint8Array([0xde, 0xad]);
    const ct = ctrEncrypt(pt, NIST_KEY, nonce);
    expect(ct.length).toBe(2); // No padding — same length as plaintext
    const result = ctrDecrypt(ct, NIST_KEY, nonce);
    expect(toHex(result)).toBe("dead");
  });

  it("encrypts empty plaintext", () => {
    const nonce = fromHex("000000000000000000000000");
    const ct = ctrEncrypt(new Uint8Array(0), NIST_KEY, nonce);
    expect(ct.length).toBe(0);
  });

  it("rejects wrong nonce length", () => {
    expect(() => ctrEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, new Uint8Array(16))).toThrow("12 bytes");
  });

  it("different nonces produce different ciphertext", () => {
    const n1 = fromHex("000000000000000000000001");
    const n2 = fromHex("000000000000000000000002");
    const ct1 = ctrEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, n1);
    const ct2 = ctrEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, n2);
    expect(toHex(ct1)).not.toBe(toHex(ct2));
  });

  it("encryption equals decryption (stream cipher property)", () => {
    const nonce = fromHex("aabbccddeeff001122334455");
    const ct = ctrEncrypt(NIST_PLAINTEXT_ALL, NIST_KEY, nonce);
    // CTR decrypt is the same function as encrypt
    const pt = ctrEncrypt(ct, NIST_KEY, nonce);
    expect(toHex(pt)).toBe(toHex(NIST_PLAINTEXT_ALL));
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// GCM Mode Tests
// ─────────────────────────────────────────────────────────────────────────────

describe("GCM Mode", () => {
  it("matches NIST GCM test vector (test case 3)", () => {
    // NIST GCM test case from the specification
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    const pt = fromHex(
      "d9313225f88406e5a55909c5aff5269a" +
      "86a7a9531534f7da2e4c303d8a318a72" +
      "1c3c0c95956809532fcf0e2449a6b525" +
      "b16aedf5aa0de657ba637b391aafd255"
    );
    const expectedCt = fromHex(
      "42831ec2217774244b7221b784d0d49c" +
      "e3aa212f2c02a4e035c17e2329aca12e" +
      "21d514b25466931c7d8f6a5aac84aa05" +
      "1ba30b396a0aac973d58e091473f5985"
    );
    const expectedTag = fromHex("4d5c2af327cd64a62cf35abd2ba6fab4");

    const { ciphertext, tag } = gcmEncrypt(pt, key, iv);
    expect(toHex(ciphertext)).toBe(toHex(expectedCt));
    expect(toHex(tag)).toBe(toHex(expectedTag));
  });

  it("round-trips with AAD", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    const pt = fromHex("d9313225f88406e5a55909c5aff5269a");
    const aad = fromHex("feedfacedeadbeeffeedfacedeadbeef");

    const { ciphertext, tag } = gcmEncrypt(pt, key, iv, aad);
    const result = gcmDecrypt(ciphertext, key, iv, aad, tag);
    expect(toHex(result)).toBe(toHex(pt));
  });

  it("rejects tampered ciphertext", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    const pt = fromHex("d9313225f88406e5a55909c5aff5269a");

    const { ciphertext, tag } = gcmEncrypt(pt, key, iv);

    // Flip a bit in the ciphertext
    const tampered = new Uint8Array(ciphertext);
    tampered[0] ^= 0x01;

    expect(() => gcmDecrypt(tampered, key, iv, new Uint8Array(0), tag)).toThrow("tag mismatch");
  });

  it("rejects tampered tag", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    const pt = fromHex("d9313225f88406e5a55909c5aff5269a");

    const { ciphertext, tag } = gcmEncrypt(pt, key, iv);

    // Flip a bit in the tag
    const tamperedTag = new Uint8Array(tag);
    tamperedTag[0] ^= 0x01;

    expect(() => gcmDecrypt(ciphertext, key, iv, new Uint8Array(0), tamperedTag)).toThrow("tag mismatch");
  });

  it("rejects wrong AAD", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    const pt = fromHex("d9313225f88406e5a55909c5aff5269a");
    const aad = fromHex("feedfacedeadbeef");

    const { ciphertext, tag } = gcmEncrypt(pt, key, iv, aad);

    // Try decrypting with different AAD
    const wrongAad = fromHex("deadbeeffeedface");
    expect(() => gcmDecrypt(ciphertext, key, iv, wrongAad, tag)).toThrow("tag mismatch");
  });

  it("encrypts empty plaintext with AAD (authentication only)", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    const aad = fromHex("feedfacedeadbeef");

    const { ciphertext, tag } = gcmEncrypt(new Uint8Array(0), key, iv, aad);
    expect(ciphertext.length).toBe(0);
    expect(tag.length).toBe(16);

    // Should decrypt successfully
    const result = gcmDecrypt(ciphertext, key, iv, aad, tag);
    expect(result.length).toBe(0);
  });

  it("rejects wrong IV length", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    expect(() => gcmEncrypt(new Uint8Array(0), key, new Uint8Array(16))).toThrow("12 bytes");
  });

  it("rejects wrong tag length", () => {
    const key = fromHex("feffe9928665731c6d6a8f9467308308");
    const iv = fromHex("cafebabefacedbaddecaf888");
    expect(() => gcmDecrypt(new Uint8Array(0), key, iv, new Uint8Array(0), new Uint8Array(8))).toThrow("16 bytes");
  });

  it("NIST GCM test case 2 (empty plaintext, empty AAD)", () => {
    // Test case 2: key, no plaintext, no AAD — tests just the tag generation
    const key = fromHex("00000000000000000000000000000000");
    const iv = fromHex("000000000000000000000000");

    const { ciphertext, tag } = gcmEncrypt(new Uint8Array(0), key, iv);
    expect(ciphertext.length).toBe(0);
    expect(toHex(tag)).toBe("58e2fccefa7e3061367f1d57a4e7455a");
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// Cross-mode Tests
// ─────────────────────────────────────────────────────────────────────────────

describe("Cross-mode behavior", () => {
  it("ECB and CBC produce different ciphertext for same plaintext", () => {
    const iv = fromHex("000102030405060708090a0b0c0d0e0f");
    const ecbCt = ecbEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY);
    const cbcCt = cbcEncrypt(NIST_PLAINTEXT_BLOCK1, NIST_KEY, iv);
    expect(toHex(ecbCt.slice(0, 16))).not.toBe(toHex(cbcCt.slice(0, 16)));
  });

  it("CTR produces same-length ciphertext as plaintext", () => {
    const nonce = fromHex("000000000000000000000000");
    for (const len of [0, 1, 15, 16, 17, 31, 32, 100]) {
      const pt = new Uint8Array(len).fill(0xaa);
      const ct = ctrEncrypt(pt, NIST_KEY, nonce);
      expect(ct.length).toBe(len);
    }
  });

  it("ECB identical blocks produce identical ciphertext (demonstrating weakness)", () => {
    // This is the ECB penguin problem: two identical plaintext blocks
    // produce identical ciphertext blocks
    const twoSameBlocks = new Uint8Array(32);
    twoSameBlocks.set(NIST_PLAINTEXT_BLOCK1, 0);
    twoSameBlocks.set(NIST_PLAINTEXT_BLOCK1, 16);
    const ct = ecbEncrypt(twoSameBlocks, NIST_KEY);
    // First and second ciphertext blocks should be identical in ECB
    expect(toHex(ct.slice(0, 16))).toBe(toHex(ct.slice(16, 32)));
  });

  it("CBC identical blocks produce different ciphertext", () => {
    // CBC chains blocks, so identical plaintext blocks produce different
    // ciphertext — this is exactly what ECB lacks
    const iv = fromHex("000102030405060708090a0b0c0d0e0f");
    const twoSameBlocks = new Uint8Array(32);
    twoSameBlocks.set(NIST_PLAINTEXT_BLOCK1, 0);
    twoSameBlocks.set(NIST_PLAINTEXT_BLOCK1, 16);
    const ct = cbcEncrypt(twoSameBlocks, NIST_KEY, iv);
    expect(toHex(ct.slice(0, 16))).not.toBe(toHex(ct.slice(16, 32)));
  });
});

File: code/packages/cpp/aes-modes/tests/aes_modes_test.cpp
// Tests for the C++ aes-modes, using the header-only iso_test.h harness (pure
// ISO). Vectors are NIST SP 800-38A (ECB/CBC) and the classic GCM test cases,
// matching the Rust crate's own tests.
#include "iso_test.h"

#include <array>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

#include "aes_modes.hpp"

namespace am = ca::aes_modes;
using Bytes = std::vector<std::uint8_t>;

static Bytes from_hex(const std::string& hex) {
    Bytes out;
    for (std::size_t i = 0; i + 1 < hex.size(); i += 2) {
        auto nib = [](char c) {
            return c <= '9' ? c - '0' : (c | 0x20) - 'a' + 10;
        };
        out.push_back(static_cast<std::uint8_t>((nib(hex[i]) << 4) |
                                                nib(hex[i + 1])));
    }
    return out;
}

int main() {
    const Bytes key = from_hex("2b7e151628aed2a6abf7158809cf4f3c");
    const Bytes pt = from_hex(
        "6bc1bee22e409f96e93d7e117393172aae2d8a571e03ac9c9eb76fac45af8e51"
        "30c81c46a35ce411e5fbc1191a0a52eff69f2445df4f9b17ad2b417be66c3710");

    // ── PKCS#7 ──────────────────────────────────────────────────────────────
    {
        Bytes aligned(16, 0xAA);
        Bytes padded = am::pkcs7_pad(aligned);
        ISO_CHECK_EQ_UINT(padded.size(), 32u);
        for (std::size_t i = 16; i < 32; ++i) {
            ISO_CHECK(padded[i] == 16);
        }
        Bytes thirteen(13, 0xBB);
        Bytes p13 = am::pkcs7_pad(thirteen);
        ISO_CHECK_EQ_UINT(p13.size(), 16u);
        ISO_CHECK(p13[13] == 3 && p13[15] == 3);
        Bytes five = {1, 2, 3, 4, 5};
        ISO_CHECK(am::pkcs7_unpad(am::pkcs7_pad(five)) == five);
        auto throws = [](auto fn) {
            try {
                fn();
            } catch (const std::invalid_argument&) {
                return true;
            }
            return false;
        };
        Bytes bad(16, 0);
        bad[15] = 0;
        ISO_CHECK(throws([&] { am::pkcs7_unpad(bad); }));
        bad[15] = 2;
        bad[14] = 3;
        ISO_CHECK(throws([&] { am::pkcs7_unpad(bad); }));
    }

    // ── ECB (NIST SP 800-38A) ───────────────────────────────────────────────
    {
        Bytes ct = am::ecb_encrypt(pt, key);
        ISO_CHECK_EQ_UINT(ct.size(), 80u);
        ISO_CHECK(Bytes(ct.begin(), ct.begin() + 16) ==
                  from_hex("3ad77bb40d7a3660a89ecaf32466ef97"));
        ISO_CHECK(Bytes(ct.begin() + 16, ct.begin() + 32) ==
                  from_hex("f5d3d58503b9699de785895a96fdbaaf"));
        ISO_CHECK(am::ecb_decrypt(ct, key) == pt);
    }

    // ── CBC (NIST SP 800-38A) ───────────────────────────────────────────────
    {
        Bytes iv = from_hex("000102030405060708090a0b0c0d0e0f");
        Bytes ct = am::cbc_encrypt(pt, key, iv);
        ISO_CHECK_EQ_UINT(ct.size(), 80u);
        ISO_CHECK(Bytes(ct.begin(), ct.begin() + 16) ==
                  from_hex("7649abac8119b246cee98e9b12e9197d"));
        ISO_CHECK(Bytes(ct.begin() + 16, ct.begin() + 32) ==
                  from_hex("5086cb9b507219ee95db113a917678b2"));
        ISO_CHECK(am::cbc_decrypt(ct, key, iv) == pt);
        bool threw = false;
        try {
            am::cbc_encrypt(pt, key, Bytes(8, 0));
        } catch (const std::invalid_argument&) {
            threw = true;
        }
        ISO_CHECK(threw);  // wrong IV length
    }

    // ── CTR (round trip; no padding) ────────────────────────────────────────
    {
        Bytes nonce = from_hex("f0f1f2f3f4f5f6f7f8f9fafb");
        Bytes ct = am::ctr_encrypt(pt, key, nonce);
        ISO_CHECK_EQ_UINT(ct.size(), pt.size());  // no padding
        ISO_CHECK(am::ctr_decrypt(ct, key, nonce) == pt);
        bool threw = false;
        try {
            am::ctr_encrypt(Bytes{0}, key, Bytes(16, 0));
        } catch (const std::invalid_argument&) {
            threw = true;
        }
        ISO_CHECK(threw);  // wrong nonce length
    }

    // ── GCM test case 3 (64-byte plaintext, empty AAD) ──────────────────────
    {
        Bytes gkey = from_hex("feffe9928665731c6d6a8f9467308308");
        Bytes iv = from_hex("cafebabefacedbaddecaf888");
        Bytes gpt = from_hex(
            "d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a72"
            "1c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b391aafd255");
        Bytes want_ct = from_hex(
            "42831ec2217774244b7221b784d0d49ce3aa212f2c02a4e035c17e2329aca12e"
            "21d514b25466931c7d8f6a5aac84aa051ba30b396a0aac973d58e091473f5985");
        Bytes want_tag_v = from_hex("4d5c2af327cd64a62cf35abd2ba6fab4");
        am::Block want_tag;
        for (std::size_t i = 0; i < 16; ++i) {
            want_tag[i] = want_tag_v[i];
        }

        auto [ct, tag] = am::gcm_encrypt(gpt, gkey, iv, Bytes{});
        ISO_CHECK(ct == want_ct);
        ISO_CHECK(tag == want_tag);
        ISO_CHECK(am::gcm_decrypt(ct, gkey, iv, Bytes{}, tag) == gpt);

        // tampered ciphertext / tag rejected.
        auto auth_fails = [&](Bytes c, am::Block t) {
            try {
                am::gcm_decrypt(c, gkey, iv, Bytes{}, t);
            } catch (const am::AuthenticationError&) {
                return true;
            }
            return false;
        };
        Bytes tampered = ct;
        tampered[0] ^= 0x01;
        ISO_CHECK(auth_fails(tampered, tag));
        am::Block bad_tag = tag;
        bad_tag[0] ^= 0x01;
        ISO_CHECK(auth_fails(ct, bad_tag));
    }

    // ── GCM empty pt / empty aad (tag known-answer) ─────────────────────────
    {
        Bytes gkey(16, 0);
        Bytes iv(12, 0);
        auto [ct, tag] = am::gcm_encrypt(Bytes{}, gkey, iv, Bytes{});
        ISO_CHECK_EQ_UINT(ct.size(), 0u);
        Bytes want = from_hex("58e2fccefa7e3061367f1d57a4e7455a");
        ISO_CHECK(Bytes(tag.begin(), tag.end()) == want);
    }

    // ── GCM with AAD round trip; wrong AAD rejected ─────────────────────────
    {
        Bytes gkey = from_hex("feffe9928665731c6d6a8f9467308308");
        Bytes iv = from_hex("cafebabefacedbaddecaf888");
        Bytes gpt = from_hex("d9313225f88406e5a55909c5aff5269a");
        Bytes aad = from_hex("feedfacedeadbeeffeedfacedeadbeef");
        auto [ct, tag] = am::gcm_encrypt(gpt, gkey, iv, aad);
        ISO_CHECK(am::gcm_decrypt(ct, gkey, iv, aad, tag) == gpt);
        bool threw = false;
        try {
            am::gcm_decrypt(ct, gkey, iv, from_hex("deadbeeffeedface"), tag);
        } catch (const am::AuthenticationError&) {
            threw = true;
        }
        ISO_CHECK(threw);
    }

    return ISO_TEST_RESULT();
}
